desc/magnetic_fields.py

```diff
@@ -795,3 +831,464 @@ def odefun(rpz, s):
+                    mode="grad",
+                ).compute,
+                in_axes=[0, 0],
+            )
+        if potential_dzeta:
+            self._potential_z = partial(potential_dzeta, **self.params)
+        else:
+            self._potential_z = vmap(
+                Derivative(
+                    partial(self.potential, **self.params), argnum=1, mode="grad"
+                ).compute,
+                in_axes=[0, 0],
+            )
+
+    def _compute_surface_current(self, surface_grid=None, params=None):
+        if surface_grid is None:
+            surface_grid = self.surface_grid
+        if (params is None) or (len(params) == 0):
+            params = self._params
+
+        # surface is the source of current density for the magnetic field
+        # compute source positions rs, and their theta/zeta derivatives
+        # store in "rpz" basis so can rotate them during compute_magnetic_field
+        data = self.compute(
+            ["x", "e_theta", "e_zeta", "|e_theta x e_zeta|"],
+            grid=surface_grid,
+            basis="xyz",
+        )
+        self._rs = xyz2rpz_vec(data["x"], phi=self.surface_grid.nodes[:, 2])
+        theta = surface_grid.nodes[:, 1]
+        zeta = surface_grid.nodes[:, 2]
+        ns_mag = data["|e_theta x e_zeta|"]
+        # compute the dV element for the surface needed for biot savart
+        self._dV = (
+            surface_grid.weights * ns_mag / surface_grid.NFP
+        )  # divide by NFP here so that the NFP loop
+        # in compute_magnetic_field is correct
+
+        # compute the potential derivatives on the surface
+        phi_t = self._potential_t(theta, zeta)
+        phi_z = self._potential_z(theta, zeta)
+
+        # compute K = n x nabla(Phi)
+        self._K = xyz2rpz_vec(
+            -(phi_t * (1 / ns_mag) * data["e_zeta"].T).T
+            + (phi_z * (1 / ns_mag) * data["e_theta"].T).T,
+            phi=self.surface_grid.nodes[:, 2],
+        )
+
+    def compute_magnetic_field(
+        self, coords, params=None, basis="rpz", surface_grid=None
+    ):
+        """Compute magnetic field at a set of points.
+
+        Parameters
+        ----------
+        coords : array-like shape(N,3) or Grid
+            cylindrical or cartesian coordinates
+        params : dict, optional
+            parameters to pass to current potential function
+        basis : {"rpz", "xyz"}
+            basis for input coordinates and returned magnetic field
+        surface_grid : Grid,
+            grid upon which to evaluate the surface current density K
+
+        Returns
+        -------
+        field : ndarray, shape(N,3)
+            magnetic field at specified points
+
+        """
+        assert basis.lower() in ["rpz", "xyz"]
+        if isinstance(coords, Grid):
+            coords = coords.nodes
+        coords = jnp.atleast_2d(coords)
+        if basis == "rpz":
+            coords = rpz2xyz(coords)
+
+        if (params is None) or (len(params) == 0):
```

Comment:
i don't think we need this here, the one in `_compute_surface_current` will do the same thing